Implement command-line arguments parser

svitbar opened the pull request:
# Purpose
To provide user with different options to interact with our application. It will enable app to recognize and process the request and execute an appropriate actions. 
# Summary
Created a help menu. The app is provided with all needed to user commands with different flags. Implemented error handling to inappropriate request.

svitbar committed:
Implement command-line argument parsing

Implement the 'ls' command.

svitbar committed:
Implement parsing the 'delete' command

svitbar committed:
Implement parsing the 'complete' command

svitbar committed:
Implement parsing the 'add' command

svitbar committed:
Implement parsing the 'edit' command

svitbar committed:
Add new dependency

svitbar committed:
Limit amount of commands to one at a time

svitbar committed:
Do some improvements

- change some description messages;
- replace id check to an appropriate command;
- remove unnecessary code.

svitbar committed:
Fix string length error

Add an object with error messages to each case.

svitbar committed:
Add an example of running function

MytsV commented on src/parser.js:
```diff
@@ -0,0 +1,143 @@
+const yargs = require('yargs/yargs');
+const {hideBin} = require('yargs/helpers');
```
This dependency is not listed in package.json. You should include package.json and package-lock.json files in your next commit for me to proceed with a review.

svitbar replied:
It seems like I installed it globally. Now I did it in appropriate way.

MytsV replied:
Thanks!

MytsV reviewed (changes requested):
Please, include dependency information in your next commit with package.json and package-lock.json files.

svitbar committed:
Install the dependency in appropriate way

MytsV commented on src/parser.js:
```diff
@@ -0,0 +1,143 @@
+const {hideBin} = require('yargs/helpers');
+const moment = require('moment');
+
+const errorMsg = {
+  ID_LENGTH: 'ID must be a string with a length of 16',
+  DATE_FMT: 'Invalid date. Provide a date in YYYY-MM-DD HH:MM:SS format.',
+  OPTION_AMOUNT: 'At least one option is required',
+  EMPTY_REQ: 'Empty request. Please input a command',
+};
```
- It's quite rare to see SCREAMING_SNAKE_CASE in names of object keys in JS. I would rather prefer camelCase.
- Always copying date format like this may be human error-prone, better put it in a constant.
```suggestion
const dateFmt = 'YYYY-MM-DD HH:MM:SS';

const errorMsg = {
  ID_LENGTH: 'ID must be a string with a length of 16',
  DATE_FMT: `Invalid date. Provide a date in ${dateFmt} format.`,
  OPTION_AMOUNT: 'At least one option is required',
  EMPTY_REQ: 'Empty request. Please input a command',
};
```
- The word 'amount' is used with something uncountable, so OPTION_AMOUNT identifier is a bit confusing. I would rather use optionCount/optionsAbsent

MytsV commented on package.json:
```diff
@@ -12,6 +12,10 @@
     "eslint-config-google": "^0.14.0",
-    "mocha": "^10.2.0"
+    "mocha": "^10.2.0",
+    "yargs": "^17.7.2"
```
yargs should be installed as a regular dependency, not a developer dependency

MytsV commented on src/parser.js:
```diff
@@ -0,0 +1,143 @@
+    .strict()
+    .demandCommand(1, errorMsg.EMPTY_REQ)
+    .help()
+    .parse();
+
+module.exports = arg;
```
We must export some structure with parse() defined on it, not the arguments themselves. Then you'll be able to:
- easily unit test the parser
- trace and handle errors in it from outside

A suggestion for restructuring the code will be in the main comment.

MytsV commented on src/parser.js:
```diff
@@ -0,0 +1,143 @@
+const arg = yargs(hideBin(process.argv))
+    .command('ls', 'Output the list of tasks', (yargs) => {
+      yargs
+          .group(['due', 'overdue', 'completed'], 'Filter Options:')
+          .option('due', {
+            describe: 'Filter tasks to show only pending tasks',
+            alias: 'd',
+            type: 'boolean',
+            conflicts: ['overdue', 'completed'],
+          })
+          .option('overdue', {
+            describe: 'Filter tasks to show only overdue tasks',
+            alias: 'o',
+            type: 'boolean',
+            conflicts: ['due', 'completed'],
+          })
+          .option('completed', {
+            describe: 'Filter tasks to show only completed tasks',
+            alias: 'c',
+            type: 'boolean',
+            conflicts: ['due', 'overdue'],
+          })
+          .strictOptions();
```
I'm sorry, but such chaining is so badly readable. Please, see my suggestions about overall structure.

MytsV commented on src/parser.js:
```diff
@@ -0,0 +1,143 @@
+        type: 'string',
+        demandOption: true,
+        coerce: (id) => {
+          if (id.length !== 16) {
+            throw new Error(errorMsg.ID_LENGTH);
+          }
+          return id;
```
- Magic number of 16
- Better to take this logic out to a separate function to avoid repetition
- Checks for ID must differ, please see my comment on the database updates

MytsV commented on src/parser.js:
```diff
@@ -0,0 +1,143 @@
+          .strictOptions();
+    }, function(arg) {
+      // TODO: run the function
+      console.log(arg);
```
My preference is to throw an Error if such functionality is unimplemented. You can trace which command is run by checking the error message.
```suggestion
      throw Error('ls called, but is not implemented yet');
```

MytsV commented on src/parser.js:
```diff
@@ -0,0 +1,143 @@
+        describe: 'task\'s ID',
+        type: 'string',
+        demandOption: true,
+        coerce: (id) => {
+          if (id.length !== 16) {
+            throw new Error(errorMsg.ID_LENGTH);
+          }
+          return id;
+        },
```
Code duplication, more reasons to make this part a separate function

MytsV commented on src/parser.js:
```diff
@@ -0,0 +1,143 @@
+            alias: 'l',
+            type: 'data',
+            demandOption: true,
+            coerce: (deadline) => {
+              const date = moment(deadline, 'YYYY-MM-DD HH:mm:ss', true);
+              if (!date.isValid()) {
+                throw new Error(errorMsg.DATE_FMT);
+              }
+              return date.format('YYYY-MM-DD HH:mm:ss');
+            },
```
- This part of the code is duplicated later and  should be decomposed
- Please, use a constant for the date format

MytsV commented on src/parser.js:
```diff
@@ -0,0 +1,143 @@
+            describe: 'task\'s hash index',
+            type: 'string',
+            demandOption: true,
+            coerce: (id) => {
+              if (id.length !== 16) {
+                throw new Error(errorMsg.ID_LENGTH);
+              }
+              return id;
+            },
```
Duplicated code fragment

MytsV commented on src/parser.js:
```diff
@@ -0,0 +1,143 @@
+            describe: 'Task\'s deadline',
+            alias: 'l',
+            type: 'date',
+            coerce: (deadline) => {
+              const date = moment(deadline, 'YYYY-MM-DD HH:mm:ss', true);
+              if (!date.isValid()) {
+                throw new Error(errorMsg.DATE_FMT);
+              }
+              return date.format('YYYY-MM-DD HH:mm:ss');
+            },
```
Duplicated code fragment

MytsV commented on src/parser.js:
```diff
@@ -0,0 +1,143 @@
+            describe: 'Description of the task',
+            alias: 'd',
+            type: 'string',
+            demandOption: true,
```
Per laboratory task, this flag should be optional

MytsV commented on src/parser.js:
```diff
@@ -0,0 +1,143 @@
+            describe: 'Task\'s deadline',
+            alias: 'l',
+            type: 'data',
+            demandOption: true,
```
One more flag that should be optional

MytsV commented on src/parser.js:
```diff
@@ -0,0 +1,143 @@
+            conflicts: ['due', 'overdue'],
+          })
+          .strictOptions();
+    }, function(arg) {
```
It's better to keep consistency with lambda functions

MytsV commented on src/parser.js:
```diff
@@ -0,0 +1,143 @@
+      yargs.positional('id', {
+        describe: 'task\'s ID',
+        type: 'string',
+        demandOption: true,
```
Per yargs documentation this option is not used for .positional function

MytsV commented on src/parser.js:
## Structure

- Excessive chaining should be fixed
- Export should be a structure with `parse()` method
- Coerce checks must be decomposed
- Date format must be a constant to avoid repetition
- We may not need errorMsg structure and use singular constants instead
Suggestion for updated structure to fix chaining (not everything from the list above is fixed here!):
```js
const yargs = require('yargs/yargs');
const {hideBin} = require('yargs/helpers');

const dateFmt = 'YYYY-MM-DD HH:MM:SS';

const errorMsg = {
  ID_LENGTH: 'ID must be a string with a length of 16',
  DATE_FMT: `Invalid date. Provide a date in ${dateFmt} format.`,
  OPTION_AMOUNT: 'At least one option is required',
  EMPTY_REQ: 'Empty request. Please input a command',
};

const parser = yargs(hideBin(process.argv));

const setLs = () => {
  parser.command('ls', 'Output the list of tasks', (yargs) => {
    yargs.group(['due', 'overdue', 'completed'], 'Filter options:');

    yargs.option('due', {
      describe: 'Filter tasks to show only pending tasks',
      alias: 'd',
      type: 'boolean',
      conflicts: ['overdue', 'completed'],
    });

    yargs.option('overdue', {
      describe: 'Filter tasks to show only overdue tasks',
      alias: 'o',
      type: 'boolean',
      conflicts: ['due', 'completed'],
    });

    yargs.option('completed', {
      describe: 'Filter tasks to show only completed tasks',
      alias: 'c',
      type: 'boolean',
      conflicts: ['due', 'overdue'],
    });

    yargs.strictOptions();
  });
};

// Decomposition example
const checkId = (id) => {
  if (id.length !== 16) {
    throw new Error(errorMsg.ID_LENGTH);
  }
  return id;
};

const setDelete = () => {
  parser.command('delete <id>', 'Delete task', (yargs) => {
    yargs.positional('id', {
      describe: 'task\'s ID',
      type: 'string',
      coerce: checkId,
    });
  });
};

/*
 I wouldn't like to expose the whole yargs structure,
 so only parse function is to be exported.
 */
const parse = () => {
  setLs();
  setDelete();
  // Add more commands :)

  // It may be this way, but.. I just don't like it
  parser
      .strict()
      .demandCommand(1, errorMsg.EMPTY_REQ)
      .help();
  /*
  I would write it this way:
  parser.strict()
  parser.demandCommand(1, errorMsg.EMPTY_REQ)
  parser.help()
   */
  return parser.parse();
};

module.exports = {parse};

```
Moreover, I don't think that we should pass a function to command! parse() should just return arguments. The code will be so much easier to test and debug this way.
## Style
Everything is fine in general, but with a few comments will be even greater ;)
## Functionality
- For 'add' command description and deadline should be optional parameters (per task)
- Sorry, I forgot to mention that in new structure of database that we discussed the IDs will be represented just as a number, not SHA1 hash. So, checks for ID should be reimplemented

MytsV reviewed (changes requested):
## File names
Maybe, cli_parser.js will be a better name, to be more specific?
## Unit tests

If you have enough time, i would request you to cover the parser with tests. It's not hard, you'll just need to stub process.argv. If you need further explanation, refer [to this answer](https://stackoverflow.com/questions/71476237/how-to-mock-or-stub-process-argv) or text me :)

svitbar committed:
Reinstall yargs dependency

svitbar committed:
Rename the file

svitbar committed:
Improve constants

- decompose constants structure;
- check syntax;
- create constant for magic number;
- create a constant for date format.

svitbar committed:
Write a function for checking id

svitbar committed:
Create function for checking date

Improve date formating.

svitbar committed:
Decompose ls part

svitbar committed:
Decompose complete part

svitbar committed:
Decompose delete part

svitbar committed:
Decompose add part

svitbar committed:
Decompose edit part

svitbar committed:
Refactoring

- solve id range issue;
- remove unnecessary call of funtion;
- remove unnecessary options.

svitbar committed:
Add checking empty fields

svitbar committed:
Remove unnecessary code

MytsV reviewed (approved):
Everything is fine, thank you for the changes!
One non-substantial detail is that we don't have to check the range of ID, because it may be changed with implementation and is anyway checked on the side of the database. But for now its OK.
As for the tests, I understand that there's very little time left, so let's trust our hands in testing :)

MytsV committed:
Merge master into barytska_wip_parser

Resolve package manager conflicts
